src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs

```diff
@@ -0,0 +1,213 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.SpaProxy
+{
+    internal class SpaProxyLaunchManager : IHostedService, IDisposable
+    {
+        private readonly SpaDevelopmentServerOptions _options;
+        private readonly ILogger<SpaProxyLaunchManager> _logger;
+        private readonly HttpClient _httpClient = new(new HttpClientHandler()
+        {
+            // It's ok for us to do this here since this service is only plugged in during development.
+            ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
+        });
+
+        private Process? _spaProcess;
+        private bool _disposedValue;
+
+        public SpaProxyLaunchManager(ILogger<SpaProxyLaunchManager> logger)
+        {
+            _options = new SpaDevelopmentServerOptions();
+            var configuration = new ConfigurationBuilder()
+                .AddJsonFile(Path.Combine(AppContext.BaseDirectory, "spa.proxy.json"))
+                .Build();
+            configuration.GetSection("SpaProxyServer").Bind(_options);
+            _logger = logger;
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Starting SPA development server");
+            var running = await ProbeSpaDevelopmentServerUrl(cancellationToken);
+            if (running)
+            {
+                _logger.LogInformation($"Found SPA development server running at {_options.ServerUrl}");
+            }
+            else
+            {
+                _logger.LogInformation($"No SPA development server running at {_options.ServerUrl} found.");
+                await StartSpaProcessAndProbeForLiveness(cancellationToken);
+            }
+        }
+
+        private async Task<bool> ProbeSpaDevelopmentServerUrl(CancellationToken cancellationToken)
+        {
+            try
+            {
+                var response = await _httpClient.GetAsync(_options.ServerUrl, cancellationToken);
+                var running = response.IsSuccessStatusCode;
+                return running;
+            }
+            catch (HttpRequestException httpException)
+            {
+                _logger.LogDebug(httpException, "Failed to reach the SPA Development proxy.");
+                return false;
+            }
+        }
+
+        private async Task StartSpaProcessAndProbeForLiveness(CancellationToken cancellationToken)
+        {
+            LaunchDevelopmentProxy();
+            var sw = Stopwatch.StartNew();
+            var livenessProbeSucceeded = false;
+            var maxTimeoutReached = false;
+            while (_spaProcess != null && !_spaProcess.HasExited && !livenessProbeSucceeded && !maxTimeoutReached)
```

Comment:
Looks like the `!livenessProbeSucceeded` part of this condition isn't needed, because it only becomes true on L79, after which we'd break out of the loop anyway.

```suggestion
            while (_spaProcess != null && !_spaProcess.HasExited && !maxTimeoutReached)
```